### PR TITLE
Add Priority Connection Queue to Worker

### PR DESCRIPTION
**Pull request by nibanks**
## Description

This build on #4279 and now adds the prioritization logic at the worker queue level. High priority operations queued on a connection now result in a connection being high priority on its worker.

## Testing

CI/CD

## Documentation

N/A

### CODE REVIEW

File: src/test/lib/DataTest.cpp [hunk 1]
```diff
@@ -3556,6 +3556,248 @@ void QuicTestOperationPriority()
...
+            for (uint8_t i = 0; i < ConnectionPriorityTestContext::NumSend; ++i) {
+                delete Streams[i];
+            }
+        }
+
+        // processing                                                             | queued
+        // [1[s0]]                                                                | []
+        // [1[s0]]                                                                | [2[s(n-1)0p, s(n-1)2p, s00, s02, ..., s(n-2)0, s(n-2)2]]
+        // [1[s0]]                                                                | [2[s(n-1)0p, s(n-1)2p, s00, s02, ..., s(n-2)0, s(n-2)2], 3[s0p, s2p]]
+        // [2[s(n-1)0p, s(n-1)2p, s00, s02, ..., s(n-2)0, s(n-2)2], 3[s0p, s2p]]  | []
```
Comment (ami-GS) [inline src/test/lib/DataTest.cpp:3701]: for some reason only `2[s(n-2)2]` is processed after `3[s2p]`. Bug? or expected?

Comment (ami-GS) [reply src/test/lib/DataTest.cpp:3701]: OK simply callback call is delayed. Splitting StreamStart/StreamSend can resolve.

File: src/test/lib/DataTest.cpp [hunk 2]
```diff
@@ -3556,6 +3553,263 @@ void QuicTestOperationPriority()
...
+            }
+        }
+
+        // processing        | queued
+        // [1[s0]]           | []
+        // [1[s0]]           | [5[s0, s2]]
+        // [1[s0]]           | [4[s0p, s2p], 5[s0, s2]]
+        // [1[s0]]           | [4[s0p, s2p], 5[s0, s2]], 3[s0, s2]]
+        // [1[s0]]           | [4[s0p, s2p], 2[s0p, s2p], 5[s0, s2], 3[s0, s2]]
+        // [1[s0, s1]]       | [4[s0p, s2p], 2[s0p, s2p], 5[s0, s2], 3[s0, s2]]
```
Comment (ami-GS) [inline src/test/lib/DataTest.cpp:3762]: Could you check whether this scenario is valid?

File: src/test/lib/DataTest.cpp [hunk 3]
```diff
@@ -3556,6 +3553,263 @@ void QuicTestOperationPriority()
...
+            }
+        }
+
+        // processing        | queued
+        // [1[s0]]           | []
+        // [1[s0]]           | [5[s0, s2]]
+        // [1[s0]]           | [4[s0p, s2p], 5[s0, s2]]
+        // [1[s0]]           | [4[s0p, s2p], 5[s0, s2]], 3[s0, s2]]
+        // [1[s0]]           | [4[s0p, s2p], 2[s0p, s2p], 5[s0, s2], 3[s0, s2]]
+        // [1[s0, s2]]       | [4[s0p, s2p], 2[s0p, s2p], 5[s0, s2], 3[s0, s2]]
```
Comment (ami-GS) [inline src/test/lib/DataTest.cpp:3762]: Could you check whether this scenario is valid? the explanation of this symbol is written around line 3639

Comment (ami-GS) [reply src/test/lib/DataTest.cpp:3762]: maybe correct

**Comment by ami-GS** [comment]
Random stall issue

**Comment by ami-GS** [comment]
QuicWorkerQueueConnection has the cause of stall. but then priority mechanism randomly fail

File: src/test/lib/DataTest.cpp [hunk 4]
```diff
@@ -3556,6 +3553,263 @@ void QuicTestOperationPriority()
...
+            ExpectedStartOrder[3] = &Stream5;
+            ExpectedStartOrder[4] = &Stream3;
+            ExpectedSendOrder[0] = &Stream1;
+            ExpectedSendOrder[1] = &Stream4;
+            ExpectedSendOrder[2] = &Stream2;
+            ExpectedSendOrder[3] = &Stream5;
+            ExpectedSendOrder[4] = &Stream3;
+
+            TEST_TRUE(memcmp(Context.StartOrder, ExpectedStartOrder, sizeof(ExpectedStartOrder)) == 0);
+            TEST_TRUE(memcmp(Context.SendOrder, ExpectedSendOrder, sizeof(ExpectedSendOrder)) == 0);
```
Comment (ami-GS) [inline src/test/lib/DataTest.cpp:3804]: Sometimes [3] and [4] are flipped. timing issue

Comment (ami-GS) [reply src/test/lib/DataTest.cpp:3804]: But how operations in FIFO queue swap.....

Comment (ami-GS) [reply src/test/lib/DataTest.cpp:3804]: Found the cause.
There is a case when [3] is in worker queue before explicitly enqueue operation by API.
Conventional enqueue doesn't re-enqueue the connection. Prioritized enqueue does.

**Comment by ami-GS** [comment]
This PR itself is ready to go.

spinquic has stall issue. Investigation in progress

**Comment by ami-GS** [comment]
RegConfig.ExecutionProfile = QUIC_EXECUTION_PROFILE_TYPE_SCAVENGER
ExecConfig.ProcessorCount = 1
1 RunThread thread

doesn't resolve the stall.

**Comment by ami-GS** [comment]
@nibanks 
MacOS doesn't hit the stall issue, I thought the difference in MacOS is the number of (platform_)worker thread, but any other differences around scheduling?

**Comment by nibanks** [comment]
macOS only uses 1 CPU / worker thread, so all the scheduling/timings will be very different

**Comment by ami-GS** [comment]
GetParam and StreamClose are not released from CxPlatEventWaitForever regardless of its prioritized enqueue or not

**Comment by ami-GS** [comment]
Connection 0x2260024C0F0 is QUEUED but not processed
```
[4]2D64.1654::2024/07/31-13:21:22.758290500 [Microsoft-Quic][conn][0x22600250130] Scheduling: 1(QUEUED)
[0]2D64.2CC0::2024/07/31-13:21:22.758291200 [Microsoft-Quic][conn][0x2260024E110] Execute: 8(API.STRM_SEND)
[0]2D64.2CC0::2024/07/31-13:21:22.758292200 [Microsoft-Quic][strm][0x2267A2C01F0] Send State: 4(FIN)
[0]2D64.2CC0::2024/07/31-13:21:22.758295000 [Microsoft-Quic][strm][0x2267A2C01F0] Send State: 2(RESET)
[0]2D64.2CC0::2024/07/31-13:21:22.758295600 [Microsoft-Quic][strm][0x2267A2C01F0] Recv State: 3(STOPPED)
[0]2D64.2CC0::2024/07/31-13:21:22.758298000 [Microsoft-Quic][conn][0x2260024E110] Execute: 7(API.STRM_START)
[0]2D64.2CC0::2024/07/31-13:21:22.758298700 [Microsoft-Quic][conn][0x2260024E110] Shutdown complete, PeerFailedToAcknowledged=0.
[0]2D64.2CC0::2024/07/31-13:21:22.758299900 [Microsoft-Quic][conn][0x2260024E110] Scheduling: 0(IDLE)
[0]2D64.2CC0::2024/07/31-13:21:22.758301200 [Microsoft-Quic][conn][0x22600250130] Scheduling: 2(PROCESSING)
[0]2D64.2CC0::2024/07/31-13:21:22.758301600 [Microsoft-Quic][wrkr][0x22679472020] QueueDelay = 5
[0]2D64.2CC0::2024/07/31-13:21:22.758302100 [Microsoft-Quic][conn][0x22600250130] Execute: 0(API.CONN_CLOSE)
[0]2D64.2CC0::2024/07/31-13:21:22.758302600 [Microsoft-Quic][conn][0x22600250130] Transport Shutdown: 18446744071562084356 (Remote=0) (QS=1)
[0]2D64.2CC0::2024/07/31-13:21:22.758303300 [Microsoft-Quic][conn][0x22600250130] Shutdown complete, PeerFailedToAcknowledged=0.
[0]2D64.2CC0::2024/07/31-13:21:22.758305800 [Microsoft-Quic][conn][0x22600250130] Unregistered from 0x22679471ED0
[0]2D64.2CC0::2024/07/31-13:21:22.758306200 [Microsoft-Quic][conn][0x22600250130] Handle closed
[0]2D64.2CC0::2024/07/31-13:21:22.758364900 [Microsoft-Quic][conn][0x22600250130] Scheduling: 0(IDLE)
[0]2D64.2CC0::2024/07/31-13:21:22.758366200 [Microsoft-Quic][wrkr][0x22679472020] IsActive = 0, Arg = 4294967295
[0]2D64.2CC0::2024/07/31-13:21:22.758366600 [Microsoft-Quic][wrkr][0x22679472020] QueueDelay = 0
[4]2D64.1654::2024/07/31-13:21:22.758383400 [Microsoft-Quic][conn][0x22600250130] Destroyed
[4]2D64.1654::2024/07/31-13:21:22.758388000 [Microsoft-Quic][conn][0x2260024E110] Scheduling: 1(QUEUED)
[4]2D64.1654::2024/07/31-13:21:22.758430100 [Microsoft-Quic][conn][0x22600250130] Created, IsServer=0, CorrelationId=13668
[4]2D64.1654::2024/07/31-13:21:22.758432700 [Microsoft-Quic][conn][0x22600250130] (SeqNum=0) New Destination CID: 0x0DBFC1F91A22C99F
[4]2D64.1654::2024/07/31-13:21:22.758433100 [Microsoft-Quic][conn][0x22600250130] Initialize complete
[4]2D64.1654::2024/07/31-13:21:22.758433600 [Microsoft-Quic][conn][0x22600250130] Assigned worker: 0x22679472020
[4]2D64.1654::2024/07/31-13:21:22.758434100 [Microsoft-Quic][conn][0x22600250130] Registered with 0x22679471ED0
[0]2D64.2CC0::2024/07/31-13:21:22.758434300 [Microsoft-Quic][wrkr][0x22679472020] IsActive = 1, Arg = 1006755057
[0]2D64.2CC0::2024/07/31-13:21:22.758435100 [Microsoft-Quic][conn][0x2260024E110] Scheduling: 2(PROCESSING)
[0]2D64.2CC0::2024/07/31-13:21:22.758435500 [Microsoft-Quic][wrkr][0x22679472020] QueueDelay = 5
[4]2D64.1654::2024/07/31-13:21:22.758435900 [Microsoft-Quic][conn][0x2260024D100] Created, IsServer=0, CorrelationId=13669
[0]2D64.2CC0::2024/07/31-13:21:22.758436100 [Microsoft-Quic][conn][0x2260024E110] Execute: 7(API.STRM_START)
[0]2D64.2CC0::2024/07/31-13:21:22.758437200 [Microsoft-Quic][strm][0x2267A2C01F0] Send State: 2(RESET)
[4]2D64.1654::2024/07/31-13:21:22.758437400 [Microsoft-Quic][conn][0x2260024D100] (SeqNum=0) New Destination CID: 0x8738CC7A0BD0952E
[0]2D64.2CC0::2024/07/31-13:21:22.758437700 [Microsoft-Quic][strm][0x2267A2C01F0] Recv State: 3(STOPPED)
[4]2D64.1654::2024/07/31-13:21:22.758437900 [Microsoft-Quic][conn][0x2260024D100] Initialize complete
[4]2D64.1654::2024/07/31-13:21:22.758438400 [Microsoft-Quic][conn][0x2260024D100] Assigned worker: 0x22679472020
[4]2D64.1654::2024/07/31-13:21:22.758438900 [Microsoft-Quic][conn][0x2260024D100] Registered with 0x22679471ED0
[0]2D64.2CC0::2024/07/31-13:21:22.758439100 [Microsoft-Quic][conn][0x2260024E110] Execute: 7(API.STRM_START)
[0]2D64.2CC0::2024/07/31-13:21:22.758439900 [Microsoft-Quic][conn][0x2260024E110] Scheduling: 0(IDLE)
[4]2D64.1654::2024/07/31-13:21:22.758440800 [Microsoft-Quic][conn][0x2260024E110] Scheduling: 1(QUEUED)
[0]2D64.2CC0::2024/07/31-13:21:22.758441000 [Microsoft-Quic][wrkr][0x22679472020] IsActive = 0, Arg = 4294967295
[0]2D64.2CC0::2024/07/31-13:21:22.758441500 [Microsoft-Quic][wrkr][0x22679472020] QueueDelay = 0
[0]2D64.2CC0::2024/07/31-13:21:22.758442400 [Microsoft-Quic][wrkr][0x22679472020] IsActive = 1, Arg = 1006755066
[4]2D64.1654::2024/07/31-13:21:22.758442500 [Microsoft-Quic][conn][0x2260024C0F0] Created, IsServer=0, CorrelationId=13670
[0]2D64.2CC0::2024/07/31-13:21:22.758442900 [Microsoft-Quic][conn][0x2260024E110] Scheduling: 2(PROCESSING)
[0]2D64.2CC0::2024/07/31-13:21:22.758443400 [Microsoft-Quic][wrkr][0x22679472020] QueueDelay = 0
[0]2D64.2CC0::2024/07/31-13:21:22.758443900 [Microsoft-Quic][conn][0x2260024E110] Execute: 15(??)
[4]2D64.1654::2024/07/31-13:21:22.758444500 [Microsoft-Quic][conn][0x2260024C0F0] (SeqNum=0) New Destination CID: 0xC1C881F8584C4250
[0]2D64.2CC0::2024/07/31-13:21:22.758444600 [Microsoft-Quic][conn][0x2260024E110] Scheduling: 0(IDLE)
[4]2D64.1654::2024/07/31-13:21:22.758445200 [Microsoft-Quic][conn][0x2260024C0F0] Initialize complete
[0]2D64.2CC0::2024/07/31-13:21:22.758445500 [Microsoft-Quic][wrkr][0x22679472020] IsActive = 0, Arg = 4294967295
[4]2D64.1654::2024/07/31-13:21:22.758445600 [Microsoft-Quic][conn][0x2260024C0F0] Assigned worker: 0x22679472020
[0]2D64.2CC0::2024/07/31-13:21:22.758446000 [Microsoft-Quic][wrkr][0x22679472020] QueueDelay = 0
[4]2D64.1654::2024/07/31-13:21:22.758446100 [Microsoft-Quic][conn][0x2260024C0F0] Registered with 0x22679471ED0
[4]2D64.1654::2024/07/31-13:21:22.758447900 [Microsoft-Quic][conn][0x2260024C0F0] Scheduling: 1(QUEUED)
[4]2D64.1654::2024/07/31-13:21:22.758484100 [Microsoft-Quic][conn][0x2260024F120] Created, IsServer=0, CorrelationId=13671
[4]2D64.1654::2024/07/31-13:21:22.758485900 [Microsoft-Quic][conn][0x2260024F120] (SeqNum=0) New Destination CID: 0x7A0605C8021674C4
[4]2D64.1654::2024/07/31-13:21:22.758486300 [Microsoft-Quic][conn][0x2260024F120] Initialize complete
```

**Comment by ami-GS** [comment]
The stall bug has gone!!!!

https://github.com/microsoft/msquic/pull/4410

**Pull request merged**